### PR TITLE
User registration and sigin/signout/signup

### PR DESCRIPTION
**Pull request by blackode**
### User Authentication API 

Prior to this change, the user authentication and letting only
authorized users to access the API is not provided.

This change implements the user registration, sign in and sign out featured API

### Things covered
--------------
- API-user sign in
- API-user sign out
- API-user registration
- API-user-auth
- token generation
- route authentication
- mixed route plugs for auth routes and no auth routes
- API for current user
- test cases for Accounts and User control

-----------------

This PR delivers the following stories.

https://www.pivotaltracker.com/story/show/158870221
https://www.pivotaltracker.com/story/show/158870194

-------------------
[deliver #158870221]
[deliver #158870194]

------

## Checklist

- [x] I have read [CONTRIBUTING.md][contributing].
- [x] My code follows the [style guidelines][contributing] of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My code does not generate any (new) [`credo`][credo] and compile-time warnings.
- [x] I have updated the documentation wherever necessary.
- [x] I have added tests to cover my changes.

[contributing]: https://github.com/aviabird/snitch/CONTRIBUTING.md
[credo]: https://github.com/rrrene/credo

------
Dear Gatekeeper,

Please make sure that the commits that will be merged or rebased into the base
branch are [formatted according to our standards][commit-style].

> The only additional requirement is that the the PR number must appear in the
> commit title in brackets: `(#XXX)`

[commit-style]: https://github.com/aviabird/snitch/blob/develop/CONTRIBUTING.md#styling

### CODE REVIEW

File: apps/snitch_api/lib/snitch_api_web/api_pipeline.ex [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    otp_app: :snitch_api,
+    module: SnitchApi.Guardian,
+    error_handler: SnitchApi.AuthErrorHandler
+
+  plug(Guardian.Plug.VerifyHeader, realm: "Bearer")
+  plug(Guardian.Plug.EnsureAuthenticated)
+  plug(Guardian.Plug.LoadResource)
+
+  # If there is a session token, restrict it to an access token and validate it
+  plug(Guardian.Plug.VerifySession, claims: %{"typ" => "access"})
```
[the remark this answers is not in the document]
Comment (blackode) [reply apps/snitch_api/lib/snitch_api_web/api_pipeline.ex]: If the user puts the `fake token` in `authorization` header and tries to get into the system, it stops him/her and checks the token and validates. That is how the definition from the Guardian.

File: apps/snitch_api/lib/snitch_api_web/controllers/user_controller.ex [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  end
+
+  def logout(conn, _params) do
+    conn
+    |> Guardian.Plug.sign_out()
+    |> put_status(204)
+    |> render("logout.json-api", data: "logged out")
+  end
+
+  def current_user(conn, _params) do
```
[the remark this answers is not in the document]
Comment (blackode) [reply apps/snitch_api/lib/snitch_api_web/controllers/user_controller.ex]: That is not for `in use`, But, we need to provide `API` for the current user.

File: apps/snitch_api/lib/snitch_api_web/api_pipeline.ex (hunk 1 above)
[the remark this answers is not in the document]
Comment (pkrawat1) [reply apps/snitch_api/lib/snitch_api_web/api_pipeline.ex]: Is this typo `typ` here ?

[the remark this answers is not in the document]
Comment (blackode) [reply apps/snitch_api/lib/snitch_api_web/api_pipeline.ex]: No. It is a  the Guardian Convention

[the remark this answers is not in the document]
Comment (pkrawat1) [reply apps/snitch_api/lib/snitch_api_web/api_pipeline.ex]: ok

**Comment by pkrawat1** [comment]
Closing as of had develop merge commits, creating another PR

**Pull request closed without merging**